include/stl2/view/subrange.hpp

```diff
@@ -12,260 +12,245 @@
 	template <class T>
-	concept bool _PairLike =
-		meta::Integral<std::tuple_size<__uncvref<T>>> &&
-		std::tuple_size<__uncvref<T>>::value == 2 &&
-		meta::Trait<std::tuple_element<0, __uncvref<T>>> &&
-		meta::Trait<std::tuple_element<1, __uncvref<T>>>;
+	concept bool _PairLike = requires(T t) {
+		typename std::tuple_size<T>;
+		requires meta::Integral<std::tuple_size<T>>;
+		requires std::tuple_size<T>::value == 2;
+		typename std::tuple_element<0, T>;
+		typename std::tuple_element<1, T>;
+		requires meta::Trait<std::tuple_element<0, T>>;
+		requires meta::Trait<std::tuple_element<1, T>>;
+		{ std::get<0>(t) } -> const std::tuple_element_t<0, T>&;
+		{ std::get<1>(t) } -> const std::tuple_element_t<1, T>&;
+	};
 
 	template <class T, class U, class V>
 	concept bool _PairLikeConvertibleTo =
-		_PairLike<T> && !Range<T> &&
-		ConvertibleTo<meta::_t<std::tuple_element<0, __uncvref<T>>>, U> &&
-		ConvertibleTo<meta::_t<std::tuple_element<1, __uncvref<T>>>, V>;
+		!Range<T> && _PairLike<__uncvref<T>> &&
+		requires(T&& t) {
+			{ std::get<0>(static_cast<T&&>(t)) } -> ConvertibleTo<U>;
+			{ std::get<1>(static_cast<T&&>(t)) } -> ConvertibleTo<V>;
+		};
 
 	template <class T, class U, class V>
 	concept bool _PairLikeConvertibleFrom =
-		_PairLike<T> && !Range<T> &&
-		Constructible<T, U, V> &&
-		ConvertibleTo<U, meta::_t<std::tuple_element<0, __uncvref<T>>>> &&
-		ConvertibleTo<V, meta::_t<std::tuple_element<1, __uncvref<T>>>>;
+		!Range<T> && Same<T, __uncvref<T>> && _PairLike<T> &&
+		Constructible<T, U, V>;
 
 	template <class T>
 	concept bool _IteratorSentinelPair =
-		_PairLike<T> && !Range<T> &&
-		Sentinel<
-			meta::_t<std::tuple_element<1, __uncvref<T>>>,
-			meta::_t<std::tuple_element<0, __uncvref<T>>>>;
+		!Range<T> && Same<T, __uncvref<T>> && _PairLike<T> &&
+		Sentinel<std::tuple_element_t<1, T>, std::tuple_element_t<0, T>>;
 
 	namespace ext {
-		template <Iterator I, Sentinel<I> S = I, bool Sized = models::SizedSentinel<S, I>>
-		struct subrange
-		: view_interface<subrange<I, S, Sized>>
-		{
-		private:
-			std::tuple<I, S> rng_ {};
-			friend subrange<I, S, true>;
-		public:
-			using iterator = I;
-			using sentinel = S;
-
-			subrange() = default;
-
-			constexpr subrange(I i, S s)
-			: rng_{i, s} {}
-
-			constexpr subrange(I i, S s, difference_type_t<I> n) requires Sized
-			: subrange{i, s} {
-				STL2_EXPECT(n == s - i);
-			}
-
-			template <ConvertibleTo<I> X, ConvertibleTo<S> Y, bool B>
-			constexpr subrange(subrange<X, Y, B> r)
-			: subrange{r.begin(), r.end()} {}
-
-			template <ConvertibleTo<I> X, ConvertibleTo<S> Y, bool B>
-			constexpr subrange(subrange<X, Y, B> r, difference_type_t<I> n)
-			requires __bool<Sized> // strange GCC bug, not yet reported
-			: subrange{r.begin(), r.end(), n} {}
-
-			template <_PairLikeConvertibleTo<I, S> PairLike>
-			constexpr subrange(PairLike r)
-			: subrange{std::get<0>(std::move(r)), std::get<1>(std::move(r))} {}
+		enum subrange_kind : bool { unsized, sized };
```

Comment:
`enum class`

Reply:
Maybe we want to make this `: byte` instead of `: bool` so that we can extend this in the future without breaking ABI.